### ISSUE TITLE
COU upgrade should be smarter when skipping upgrading nova-compute

### ISSUE THREAD

**Issue by samuelallan72** [canonical/charmed-openstack-upgrader#481]
See also https://github.com/canonical/charmed-openstack-upgrader/issues/480

It's confusing that COU is happy to skip upgrading nova-compute (the hypervisors) when they are non-empty, even though future steps (such as ceph-osd) require upgraded hypervisors.

COU could be smarter, and for this situation exit at the plan stage with an error like 'cannot plan for upgrading ceph-osd because nova-compute must be upgraded first; nova-compute cannot be upgraded because they are non-empty'.  (Just a suggestion; there may be a better way to handle this situation.)

**Comment by samuelallan72** [comment]
I think this topic needs to be specced out a bit more before we can work on it. Any fix here I think is going to involve major refactoring... and there are many combinations to consider.

Should COU simply bail out if hypervisors are going to be skipped when the upgrade group is data-plane or complete cloud? Should we add a simply warning to the pre-plan cloud verification? How does this work with a partial upgrade (some nova computes skipped, but not all)? 

Thoughts on the topic? @jneo8 @chanchiwai-ray @rgildein @gabrielcocenza

**Comment by rgildein** [comment]
I remember that we decide to not check it during analyze, since analyze should only gather information about model/applications and not execute anything on top of any application. The plan phase should create a plan based on analyze result, so ideally no juju command.

If you collect such information during analyze, you can use it to verify it and provide some warning message even before plan, however you still can have plan saying that it will upgrade some nova-computes units and at the end they will be skipped, because some VMs were created on them during upgrading other applications. Potential another situation is that you skip some nova-computes, but they are in process to remove/migrate VMs, and they will finish it before it's their turn to be upgraded. What I'm trying to say is that the process of upgrade is a long process, and we cannot grant that plan will always work if something failed during it.

Plan is explicitly saying `Verify that unit 'nova-compute/x' has no VMs running`, so user is aware that we are checking it + the next step is marked with `├── ` what represent that such step is dependent on previous one (without such symbol) and will be skipped. Then there is `Verify that the workload of 'nova-compute' has been upgraded on units: nova-compute/x, nova-compute/x, nova-compute/x, nova-compute/x`, which will failed and whole upgrade stop. If you think that it's not clear, I will suggest providing some examples and better explanation in docs, but IMO making COU smarter it's not good idea here.

**Comment by rgildein** [comment]
Actually, I forgot that we are doing filtering non-empty hypervisors during plan phase. But, I still think that if something needs to fixed / improved, it should be the documentation.

**Comment by samuelallan72** [comment]
> Plan is explicitly saying Verify that unit 'nova-compute/x' has no VMs running, so user is aware that we are checking it...

@rgildein I'm not worried about that part; that is nice and clear. :+1: 

The main thing here I flagged is:

> It's confusing that COU is happy to skip upgrading nova-compute (the hypervisors) when they are non-empty, even though future steps (such as ceph-osd) require upgraded hypervisors.

It's probably low priority, but if we consider this situation:

- all hypervisors have VMs
- the user runs `cou upgrade`
  - no hypervisors are upgraded
  - when cou reaches ceph-osd, it bails out, because it only allows upgrading ceph-osd after nova-compute (but all nova-compute units were skipped).

So basically it generated an uncompletable plan.

I agree that it's probably not worth the effort improving this, but instead working on improving the docs with more examples of fully upgrading production clouds.

On the code side, I strongly think we should consider:

- removing the possibility for a bare `cou upgrade` - ie. force the user to upgrade in groups (on a production cloud, no-one is going to be using `cou upgrade`)
- reworking the upgrade groups to make it simpler

My proposal here would be to close off this ticket as not worth the effort (especially since the warning message on skipping hypervisors is clearer now too https://github.com/canonical/charmed-openstack-upgrader/pull/599 ), and discuss other options such as improving the docs and rethinking upgrade groups.

**Comment by samuelallan72** [comment]
We chatted about this in the standup today, and the conclusion was that it's ok to close this ticket, and instead I proposed we redirect the effort into improving the docs around upgrading. I opened https://github.com/canonical/charmed-openstack-upgrader/issues/601 to get it started - @rgildein and others, please feel free to open more docs tickets and/or comment on https://github.com/canonical/charmed-openstack-upgrader/issues/601 . :)
